// routing-controllers-openapi/build/generateSpec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.expressToOpenAPIPath = exports.getTags = exports.getSummary = exports.getSpec = exports.getResponses = exports.getStatusCode = exports.getContentType = exports.getRequestBody = exports.getQueryParams = exports.getPathParams = exports.getHeaderParams = exports.getPaths = exports.getOperationId = exports.getOperation = exports.getFullPath = exports.getFullExpressPath = void 0;
const tslib_1 = require("tslib");
const lodash_merge_1 = tslib_1.__importDefault(require("lodash.merge"));
const lodash_capitalize_1 = tslib_1.__importDefault(require("lodash.capitalize"));
const lodash_startcase_1 = tslib_1.__importDefault(require("lodash.startcase"));
const oa = tslib_1.__importStar(require("openapi3-ts"));
const pathToRegexp = tslib_1.__importStar(require("path-to-regexp"));
require("reflect-metadata");
const decorators_1 = require("./decorators");
function getFullExpressPath(route) {
    const { action, controller, options } = route;
    return ((options.routePrefix || '') +
        (controller.route || '') +
        (action.route || ''));
}
exports.getFullExpressPath = getFullExpressPath;
function getFullPath(route) {
    return expressToOpenAPIPath(getFullExpressPath(route));
}
exports.getFullPath = getFullPath;
function getOperation(route, schemas) {
    const operation = {
        operationId: getOperationId(route),
        parameters: [
            ...getHeaderParams(route),
            ...getPathParams(route),
            ...getQueryParams(route, schemas),
        ],
        requestBody: getRequestBody(route) || undefined,
        responses: getResponses(route),
        summary: getSummary(route),
        tags: getTags(route),
    };
    const cleanedOperation = Object.entries(operation)
        .filter(([_, value]) => value && (value.length || Object.keys(value).length))
        .reduce((acc, [key, value]) => {
        acc[key] = value;
        return acc;
    }, {});
    return (0, decorators_1.applyOpenAPIDecorator)(cleanedOperation, route);
}
exports.getOperation = getOperation;
function getOperationId(route) {
    return `${route.action.target.name}.${route.action.method}`;
}
exports.getOperationId = getOperationId;
function getPaths(routes, schemas) {
    const routePaths = routes.map((route) => ({
        [getFullPath(route)]: {
            [route.action.type]: getOperation(route, schemas),
        },
    }));
    return (0, lodash_merge_1.default)(...routePaths);
}
exports.getPaths = getPaths;
function getHeaderParams(route) {
    const headers = route.params
        .filter((p) => p.type === 'header')
        .map((headerMeta) => {
        const schema = getParamSchema(headerMeta);
        return {
            in: 'header',
            name: headerMeta.name || '',
            required: isRequired(headerMeta, route),
            schema,
        };
    });
    const headersMeta = route.params.find((p) => p.type === 'headers');
    if (headersMeta) {
        const schema = getParamSchema(headersMeta);
        headers.push({
            in: 'header',
            name: schema.$ref.split('/').pop() || '',
            required: isRequired(headersMeta, route),
            schema,
        });
    }
    return headers;
}
exports.getHeaderParams = getHeaderParams;
function getPathParams(route) {
    const path = getFullExpressPath(route);
    const tokens = pathToRegexp.parse(path);
    return tokens
        .filter((token) => token && typeof token === 'object')
        .map((token) => {
        const name = token.name + '';
        const param = {
            in: 'path',
            name,
            required: token.modifier !== '?',
            schema: { type: 'string' },
        };
        if (token.pattern && token.pattern !== '[^\\/]+?') {
            param.schema = { pattern: token.pattern, type: 'string' };
        }
        const meta = route.params.find((p) => p.name === name && p.type === 'param');
        if (meta) {
            const metaSchema = getParamSchema(meta);
            param.schema =
                'type' in metaSchema ? Object.assign(Object.assign({}, param.schema), metaSchema) : metaSchema;
        }
        return param;
    });
}
exports.getPathParams = getPathParams;
function getQueryParams(route, schemas) {
    var _a;
    const queries = route.params
        .filter((p) => p.type === 'query')
        .map((queryMeta) => {
        const schema = getParamSchema(queryMeta);
        return {
            in: 'query',
            name: queryMeta.name || '',
            required: isRequired(queryMeta, route),
            schema,
        };
    });
    const queriesMeta = route.params.find((p) => p.type === 'queries');
    if (queriesMeta) {
        const paramSchema = getParamSchema(queriesMeta);
        const paramSchemaName = paramSchema.$ref.split('/').pop() || '';
        const currentSchema = schemas[paramSchemaName];
        if (oa.isSchemaObject(currentSchema)) {
            for (const [name, schema] of Object.entries((currentSchema === null || currentSchema === void 0 ? void 0 : currentSchema.properties) || {})) {
                queries.push({
                    in: 'query',
                    name,
                    required: (_a = currentSchema.required) === null || _a === void 0 ? void 0 : _a.includes(name),
                    schema,
                });
            }
        }
    }
    return queries;
}
exports.getQueryParams = getQueryParams;
function getNamedParamSchema(param) {
    const { type } = param;
    if (type === 'file') {
        return { type: 'string', format: 'binary' };
    }
    if (type === 'files') {
        return {
            type: 'array',
            items: {
                type: 'string',
                format: 'binary',
            },
        };
    }
    return getParamSchema(param);
}
function getRequestBody(route) {
    const bodyParamMetas = route.params.filter((d) => d.type === 'body-param');
    const uploadFileMetas = route.params.filter((d) => ['file', 'files'].includes(d.type));
    const namedParamMetas = [...bodyParamMetas, ...uploadFileMetas];
    const namedParamsSchema = namedParamMetas.length > 0
        ? namedParamMetas.reduce((acc, d) => (Object.assign(Object.assign({}, acc), { properties: Object.assign(Object.assign({}, acc.properties), { [d.name]: getNamedParamSchema(d) }), required: isRequired(d, route)
                ? [...(acc.required || []), d.name]
                : acc.required })), { properties: {}, required: [], type: 'object' })
        : null;
    const contentType = uploadFileMetas.length > 0 ? 'multipart/form-data' : 'application/json';
    const bodyMeta = route.params.find((d) => d.type === 'body');
    if (bodyMeta) {
        const bodySchema = getParamSchema(bodyMeta);
        const items = 'items' in bodySchema && bodySchema.items ? bodySchema.items : bodySchema;
        const $ref = oa.isReferenceObject(items) ? items.$ref : '';
        return {
            content: {
                [contentType]: {
                    schema: namedParamsSchema
                        ? { allOf: [bodySchema, namedParamsSchema] }
                        : bodySchema,
                },
            },
            description: ($ref || '').split('/').pop(),
            required: isRequired(bodyMeta, route),
        };
    }
    else if (namedParamsSchema) {
        return {
            content: { [contentType]: { schema: namedParamsSchema } },
        };
    }
}
exports.getRequestBody = getRequestBody;
function getContentType(route) {
    const defaultContentType = route.controller.type === 'json'
        ? 'application/json'
        : 'text/html; charset=utf-8';
    const contentMeta = route.responseHandlers.find((h) => h.type === 'content-type');
    return contentMeta ? contentMeta.value : defaultContentType;
}
exports.getContentType = getContentType;
function getStatusCode(route) {
    const successMeta = route.responseHandlers.find((h) => h.type === 'success-code');
    return successMeta ? successMeta.value + '' : '200';
}
exports.getStatusCode = getStatusCode;
function getResponses(route) {
    const contentType = getContentType(route);
    const successStatus = getStatusCode(route);
    return {
        [successStatus]: {
            content: { [contentType]: {} },
            description: 'Successful response',
        },
    };
}
exports.getResponses = getResponses;
function getSpec(routes, schemas) {
    return {
        components: { schemas: {} },
        info: { title: '', version: '1.0.0' },
        openapi: '3.0.0',
        paths: getPaths(routes, schemas),
    };
}
exports.getSpec = getSpec;
function getSummary(route) {
    return (0, lodash_capitalize_1.default)((0, lodash_startcase_1.default)(route.action.method));
}
exports.getSummary = getSummary;
function getTags(route) {
    return [(0, lodash_startcase_1.default)(route.controller.target.name.replace(/Controller$/, ''))];
}
exports.getTags = getTags;
function expressToOpenAPIPath(expressPath) {
    const tokens = pathToRegexp.parse(expressPath);
    return tokens
        .map((d) => (typeof d === 'string' ? d : `${d.prefix}{${d.name}}`))
        .join('');
}
exports.expressToOpenAPIPath = expressToOpenAPIPath;
function isRequired(meta, route) {
    var _a, _b, _c;
    const globalRequired = (_c = (_b = (_a = route.options) === null || _a === void 0 ? void 0 : _a.defaults) === null || _b === void 0 ? void 0 : _b.paramOptions) === null || _c === void 0 ? void 0 : _c.required;
    return globalRequired ? meta.required !== false : !!meta.required;
}
function getParamSchema(param) {
    const { explicitType, index, object, method } = param;
    const type = Reflect.getMetadata('design:paramtypes', object, method)[index];
    if (typeof type === 'function' && type.name === 'Array') {
        const items = explicitType
            ? { $ref: '#/components/schemas/' + explicitType.name }
            : { type: 'object' };
        return { items, type: 'array' };
    }
    if (explicitType) {
        return { $ref: '#/components/schemas/' + explicitType.name };
    }
    if (typeof type === 'function') {
        if (type.prototype === String.prototype ||
            type.prototype === Symbol.prototype) {
            return { type: 'string' };
        }
        else if (type.prototype === Number.prototype) {
            return { type: 'number' };
        }
        else if (type.prototype === Boolean.prototype) {
            return { type: 'boolean' };
        }
        else if (type.name !== 'Object') {
            return { $ref: '#/components/schemas/' + type.name };
        }
    }
    return {};
}
